### PR TITLE
Refactors chemistry, adds all Elements

### PR DESCRIPTION
**Pull request by jrslagle**
Added all elements up to 95 - Americium. Names, symbols, masses, and atomic number as a comment.

Put all elements in one file. Calling one is like elements.Carbon().

Added the mass and density to all chemicals, and added a couple chemicals. 

Moved equipment entities from source/ to source/entities/

Verified that build_workflow and demo work. Although it looks like someone intentionally disabled the hold action in source/entities/avatar/controller.py:124

### CODE REVIEW

**Comment by jrslagle** [comment]
I'm eating my own words about no dead code. Are we ever going to call element 66 Dysprosium? But at least it's a complete set now.

**Comment by jrslagle** [comment]
I'm by no means finished with #46. But as much as the code base has been moving tectonically lately, I'm getting a dam submit in before I'm left behind again. This may also satisfy #38, since that was really about creating classes for all the elements.

**Comment by Master-Foo** [comment]
I pulled the code from your repo. It does not seem to reflect the changes you are submitting here. I tried to check it, but cant. So... I'm going to accept. But, if this breaks master I'm going to be pissed off.

Anyway, from what you told me, you kinda cowboyed it up and did some things I told you not to do. Like put everything in one file. So, I'll accept, but you are going to have to go back and fix it.

You can put includes in the __init__.py file to create the illusion that all files in one directory are part of a single package. Break up the files and start there. Also, in the issue description, I said to put an info dump in comments for each element make sure you do that so we have it when we need it.

**Comment by Master-Foo** [comment]
Another thing, the Abstract package is only for classes which are only ever **inherited** for instance, Animate Entity is not something you'd use in the game itself. It is just something that something in the game might inherit from. So, putting the elements and chemicals and minerals and such in Abstract is not where it belongs.

**Comment by jrslagle** [comment]
There's a lot more cowboy where that came from… 
Maybe you're just used to ordering virgins around, but this is a collaboration. 

But my purpose here is to develop my coding and collaboration skills while having fun. If you want me to make a huge structural change, convince me it's a good idea, and you'll have taught me how to fish. 

Yea you're right, chemistry/ shouldn't be in abstract/. Would systems/ be a good home? 

As far as elements.py, what are the advantages to putting each element in its own file, because all I see are many disadvantages. The advantages to my single file system are

Maintenance 
-I don't have to open and close 95 files every time I want to inspect or edit. 
-I can just scroll to check for consistency. 
-On the file tree, I can't get the element files to sort by atomic number, since file names starting with numbers aren't allowed. 
-if you want to find an element by name or symbol, just Ctrl+F. 

Usage 
-I only have to import one file, regardless of how many elements I'm using. 
-”elements.Chromium()” is more elegant than “chromium.Chromium()”

My reason for leaving the atomic number out of memory is I don't plan on ever needing to call it from code. But above each element is a comment with the element number, and they're all in numerical order, so you can easily find an element by number. What did you mean by “for bookkeeping purposes”, because I've satisfied every need I can think of. 

Instead of pasting the whole element data sheet under each element, I put a link to our source at the top of the elements list. I don't want heaping boneyards of dead code lying around distracting and daunting us. Keep it clean. Your source website is easy to navigate and consistent enough to web scrape if we want to. Besides, my design rests on the properties of chemicals and reactions as its basement layer, making elements just a novelty for error checking molecular formulas and molar masses - neither of which are required since chemicals and reactions have fixed data members. So I don't expect to add more parameters to elements very much at all.

The bigger reason elements are not required is because your "ElementalStorageUnit" is a spherical cow that makes me cringe. In the real world you rarely store or handle individual elements. You handle compounds. Oxygen, hydrogen, and chlorine are O2, H2, and Cl2 respectively. This is because pure elements, when forced to be that way, are very high energy states and will collapse into compounds the first chance they get (making them pretty explosive). Any pure elements you do see (like Aluminum or Copper) can be represented in our system as chemicals with one member atom. And only Copper the chemical will have a density and color and other properties. I'm going along with your ElementalStorageUnit for now because "start stupid, then improve" is a good design philosophy.

**Comment by jrslagle** [comment]
And once I start switching over to realistic thermodynamics, you'll see why isolated elements are stupifyingly expensive.

**Comment by Master-Foo** [comment]
This **is** a collaboration. That's why it's really important to put things where they belong. So everyone automatically knows where it is. Sure, knowing where something belongs requires experience. But typically, in MVC systems everything has a correct place and once you understand MVC systems, you know **exactly** how to do it and where to do it. This rule applies to breaking files up as well. Even Cowboys need to conform in MVC architecture.

I'm not really mad or anything. I'm mostly just giving you a bad time so you know not to do that. Mostly, I just expect everyone to follow convention and clean up when things are not where they belong. That's not unrealistic. Hey, I'll admit to being  a hypocrite. Right now I too am doing some cleanup and putting things where they belong. It's just something that needs to be done.

For instance, I followed your line of reasoning in the Joystick and Keyboard module. It's a cluster fuck in there. Now I need to go in and clean it up. Hopefully, after I'm done, I can close that book and never open it again.

As for leaving out atomic number and such, We do need it. It functions as an ID for when we use them in data structures and saving them in databases. Thats what I mean by "book keeping". You know my big thing is "Refrain from unnecessary complexity". So, I wouldn't make it a requirement if I didn't see a need for it. Maybe we don't need Elemental Abundances immediately, but it is going to be required soon. I'd rather fill out the details on what we know we are going to need and close the book on elements and hopefully never have to open it.

As for the Elemental Storage Unit being an "Spherical Cow" yes, it is. Intentionally so. We can turn this spherical cow into a cow shaped cow. But to get things working I need to keep it simple. Why? Because I have 1000 other animals to work on and a spherical cow is better than a cow shaped cow which never gets built because of complexity.

These are evolving systems. It's OK to start out simplistic and somewhat inaccurate. For instance, the ESU isn't really meant to store individual Elements, so much as a representation of quantity. If you want to spend the time to do better book keeping in those systems, that's the reason I built them. It gives us something VERY simplistic to start out and which can be improved later. Or you can do it now. But if I'm going to do it, it's going to be when it becomes a priority, which is a long way off. Because I have to create a spherical donkey and a spherical cat and a spherical chicken before I give a fuck about an accurately shaped cow. Right?

Worrying too much about making these systems completely scientifically accurate is going to make MVP impossible. We can make them scientifically accurate in the future. But scientific accuracy is something we should evolve into. Not expect from the beginning.

Trust me. I want scientific accuracy as much as you do. It's a **goal**, but we have only just begun this marathon and we need to worry about running the next kilometer. We'll get to the goal after many many kilometers. But right now we are in this kilometer.

**Comment by Master-Foo** [comment]
If you want to turn the chemistry modules into a cow shaped cow right now, create some issues and we'll discuss how to do that. Some explanations should be made more clear. For instance, it's not expected that any given object has 1,000,000,000,000,000,000 Oxygen objects instantiated to represent EACH atom. It's just a book keeping mechanism to say hey, there is a proportinal amount of Oxygen to Hydrogen in this thing and here is 1 or 2 Oxygen atom objects to reference.

**Comment by Master-Foo** [comment]
Oh, I forgot to answer your question about where `chemistry` should go. I think it's a good idea to packaged all these related classes up into 1 chemistry package.

The palace for that isn't systems because that's for classes that act as containers which have systems which communicate with each other. For instance, a kilometer contains 100 hectares and kilometer gives a structure that allows hectares to "talk" to each other and the hectare's parent kilometer.

As I said earlier, `Abstract` isn't correct either. Because that's for classes that are only ever inherited.

For now, it should go where it was. in the source directory. in the future, when / if we have more science libraries we'll probably make a standalone science library containing chemistry. But, we'll evolve into that when it is time.

**Comment by Master-Foo** [comment]
@jrslagle how close are you to pushing a cleaned version of the problems we discussed? I'm almost finished with the MVP workflow, but I need all the Chemistry pieces in place where they belong before I can finish the last couple of pieces.

We are ALMOST to the point where we can pull a mineral out of the ground, put it in the refinery and shit out a solar panel. I can probably finish it up in a couple of hours if the chemistry package is in place.

**Comment by jrslagle** [comment]
I've moved chemistry back into source.
Added atomic number.
I'm still not at all a fan of putting elements in separate files.
I'm rewriting a clone of the build workflow script, but I'm grinding to a halt.

Actually today I've been working on an overarching game design, so that we have notes to compare at our in-person meeting. I've come to my own opinions on camera type, survival, and death for instance. Been working on Market structure last. I'm up for discussion on all of it.

You're right though, we are close to MVP. I'll get back to converting the build workflow to real world units.

**Comment by Master-Foo** [comment]
Cool, all I really need is a clean directory structure. The logic of your last push was fine. That's why I accepted it. So, just focus on putting everything where it belongs and I'll be able to work with it.

I'll explain why putting elements in separate files is worth while when we meet. It only seems like a bad idea when you don't know why it's a good idea. For now, just trust my experience. I wouldn't be asking for this if it weren't important. It's not like I gain anything by sending you off on a wild goose chase, Right?

**Comment by Master-Foo** [comment]
If it helps, keep the elements in one file. Add the few things that need to be added, like the atomic number and elemental abundances, split the files up and hopefully, we never need to worry about elements ever again.

**Comment by jrslagle** [comment]
Oh yea, I argue that elemental abundances shouldn't be properties of elements. They instead should be "knowledge" of the kilometer, and broken up from there.

**Comment by jrslagle** [comment]
Why would an element care how common it is? A kilometer (or rather its terrain generator) does care how common it's going to make iron for instance.

**Comment by Master-Foo** [comment]
The Element Class just acts as a sort of Encyclopedia which we can query information about it. It doesn't really do anything useful except for provide information about itself. Perhaps in the future it might have a more complicated role. But for now it just serves as an in-game wikipedia article. Objects that use "Oxygen" simply ask the Oxygen class for information about itself.

So, when we are determining elemental abundances on various planets and determining how rare something is, kilometer can ask "Oxygen" for a **baseline** to start with, then kilometer will have a modifier for that baseline to determine it's actual content.

Different planets, kilometers, hectares, tiles will have varying quantities of any given Element. But that quantity is always going to be derived procedurally from a common baseline. The best place to put that baseline is in the Element itself.

**Comment by jrslagle** [comment]
I don't like it at all. 

The way I'd generate a random planet is to make plasma clouds of [igneous, metamorphic, sedimentary], and then within those, do sub clouds of rock types like sandstone, shale, bauxite. and then use procedural templates to lay out the minerals in those rocks. You want baseline values? Just guess to start or look them up.

Later we'll upgrade to Martian satellite infrared spectroscopy which actually gave us maps of which minerals are where. That's how we know hematite is on the surface which typically forms in water.

Via the principle of least knowledge, I'd leave elements out of the business of knowing what a given celestial body is made of. When you want to add a new planet, do you edit the abundance in those 95 files, or do you just tell the terrain generator to do the work?

**Comment by Master-Foo** [comment]
We are going to be using solid noise to generate modifiers. (Similar to plasma clouds)

Jimmy, I have a plan. It's well thought out. I'm willing to lead by example and show you why it's a good idea, but if I spend time arguing with you about every last detail that's time I can't spend working on more important things.

So, here's what we can do. We can recognize that my plan up to this point has worked out pretty well and trust that I'm moving in the right direction and that I'm requesting these features for a good reason.

Or, you can do it your way, which would actually require that you have the skill to pull it off.

Or, you can not do it. I'm not going to twist your arm. But I'm done arguing. Arguing doesn't accomplish anything. But putting the elemental abundances in the Elements will accomplish something.

**Comment by jrslagle** [comment]
Ok you're right. I'll get back to work.

**Comment by jrslagle** [comment]
No that's fair. I'm not upset. I've been trying to limit what I talk about to talk that leads to more code - since code is the only thing that moves us forward. And you're right. While you've been solving about 4 issues in a night, I'm struggling on one for 3 days. (Why won't this let me pass an array into a constructor??) So I'd totally implement these complex features if I could. But I'm very novice with python.

**Comment by jrslagle** [comment]
I'm trying
aluminas = composition.Composition([ ... ])

and in Composition

def __init__(self, *comp):

and it's telling me I'm passing an unexpected argument.

**Comment by Master-Foo** [comment]
Here's the way I look at it. I don't submit an issue I'm not reasonably certain I can't do myself. And I don't assign an issue I'm not reasonably certain that person can't do.

I don't mind when someone submits an issue they can't do. But I do mind when they expect someone else to do something they can't do. Which is why I was against Alex's idea of having a "Game Designer". We don't need "Pointy Haired Bosses". If everyone does the work they **can** do. Everyone will get **exactly** the game they deserve 100%.

As far as I'm concerned this project is a Meritocracy. I don't really consider myself the "leader". I just happen to do a lot of things which can be **confused** with leading. If anyone wants to step up and call themselves "leader" push code!

As for your problem. Push it to a branch on your Github and make an issue there. I'll try to figure it out if you can't.

**Comment by jrslagle** [comment]
Fixed it! You apparently can't have two __init___() methods with different argument profiles. You could do that stuff in Java.

**Comment by Master-Foo** [comment]
You can do something similar like this
```python
class Foo:
    def __init__(self, a, b = none):
        pass
```
then...
```python
foo0 = Foo(value0)
foo1 = Foo(value0, value1)
```

**Comment by jrslagle** [comment]
@Master-Foo (just suggesting) What about adding a baseline Composition() to the definition of each Planet()?

**Comment by Master-Foo** [comment]
The problem with that is there are literally thousands???, millions???, billions??? of possible permutations of compositions. While only 90 natural elements. So, it is more practical to make a baseline for the elements and statistically determine which composition is most likely to be pulled from that land.

Not only that, but we have hard reliable data that we already have to tell us rarity of Hydrogen VS Gold. We really don't have any starting point baseline for mineral composition.

For instance, we know that because Uranium is more rare compared to Iron, that a Uranium Oxide is going to be rare and an Iron Oxide is going to be plentiful. But to work top-down, like you are suggesting assumes we have information for a baseline which we don't have and would be a pain in the ass to gather.

You complain about putting in elemental abundances for 90 Elements. OK, now apply that logic to Gazillions of permutations of compositions.

**Comment by Master-Foo** [comment]
TLDR; The bookkeeping footprint for 90 Elements is less than the footprint of compositions.

**Comment by jrslagle** [comment]
No no. I meant a baseline Composition() of the 95 elements in the Planet(). Leave minerals out.

**Comment by jrslagle** [comment]
A given planet has a pretty characteristic Elemental composition because of how solar wind stratifies a forming solar system.

**Comment by Master-Foo** [comment]
Let's talk about it in person. I'm not entirely sure where you are going with this.
